### PR TITLE
Fix the MethodCallStatusException payload

### PR DESCRIPTION
**Pull request by dacolgit**
fix bug [https://msazure.visualstudio.com/One/_workitems/edit/13250556](https://msazure.visualstudio.com/One/_workitems/edit/13250556)

### CODE REVIEW

File: common/src/Microsoft.Azure.IIoT.Hub.Module.Framework/src/Hosting/MethodRouter.cs [hunk 1]
```diff
@@ -335,6 +342,22 @@ private class DefaultFilter : ExceptionFilterAttribute {
...
+            /// </summary>
+            /// <param name="ex"></param>
+            private MethodCallStatusExceptionModel CreateExceptionResponse(Exception ex) {
+                MethodCallStatusExceptionModel exceptionResponse = null;
+                if (!string.IsNullOrEmpty(ex.Message)) {
+                    exceptionResponse = new MethodCallStatusExceptionModel {
+                        Message = ex.Message.Split(":").First(),
+                        Description = ex.Message.Split(":", 2).Last(),
+                    };
+                }
```
Comment (karok2m) [inline common/src/Microsoft.Azure.IIoT.Hub.Module.Framework/src/Hosting/MethodRouter.cs]: This logic relies heavily on the structure of the message that is generated in the constructor of `MethodCallStatusException`, which makes me think that this should happen immediately there instead of requiring a post-processing step here:

```
$"Response {result} {errorMessage ?? ""}: {responsePayload}"
```

I think that it would be better if in the constructor itself we create the payload structure that we are trying to add here. It would also be more maintainable i the long run since it would not require knowledge of the fact that the final message is constructed in two steps, in the constructor of `MethodCallStatusException` and here.

I am not sure, but I also think that this logic might not process the case when `MethodCallStatusException` is constructed without a response payload. In that case internally the constructor that accepts response payload is called with `{}` as payload. It seems like that might not be parsed correctly here. So we need to add tests that make sure that in both cases resulting `ResponsePayload` would be a valid JSON that can be deserialized back to an object from string. And on a side note, `null` might be a better value to pass from the first constructor to the one with response payload instead of `{}`.

Comment (dacolgit) [reply common/src/Microsoft.Azure.IIoT.Hub.Module.Framework/src/Hosting/MethodRouter.cs]: I moved the creation of json paylosd in  `MethodCallStatusException`

File: common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs [hunk 2]
```diff
@@ -39,7 +39,7 @@ public class MethodCallStatusException : MethodCallException {
         /// <param name="errorMessage"></param>
         public MethodCallStatusException(string responsePayload, int result,
             string errorMessage = null) :
-            base($"Response {result} {errorMessage ?? ""}: {responsePayload}") {
+            base($" {{ \"Message\":\"Response {result} {errorMessage ?? ""}\",\"Description\": {responsePayload} }}") {
```
Comment (karok2m) [inline common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs]: Please add tests that would validate that we can reserialize `ResponsePayload` as a proper JSON afterwards. The tests should also check the cases where passed `responsePayload` or `errorMessage` are null. In both cases we should still be able to deserialize. I think that the test should also make sure that message constructed from the other constructor above is also deserializable.

Comment (dacolgit) [reply common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs]: Added unit test

File: common/src/Microsoft.Azure.IIoT.Hub.Module.Framework/src/Hosting/MethodRouter.cs [hunk 3]
```diff
@@ -300,8 +300,9 @@ private class DefaultFilter : ExceptionFilterAttribute {
                         }
                         _logger.Verbose(ex, "Method call error");
                         ex = _ef.Filter(ex, out var status);
+
                         if (ex is MethodCallStatusException) {
-                            throw ex;
+                            throw new MethodCallStatusException(ex.Message, status);
```
Comment (karok2m) [inline common/src/Microsoft.Azure.IIoT.Hub.Module.Framework/src/Hosting/MethodRouter.cs]: Wouldn't `ex` already contain the `status` ? Looks like we extract `status` from `ex` itself a few lines above.

Comment (karok2m) [reply common/src/Microsoft.Azure.IIoT.Hub.Module.Framework/src/Hosting/MethodRouter.cs]: Same comment for changes on line `331` below.

File: common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs [hunk 4]
```diff
@@ -39,7 +39,7 @@ public class MethodCallStatusException : MethodCallException {
         /// <param name="errorMessage"></param>
         public MethodCallStatusException(string responsePayload, int result,
             string errorMessage = null) :
-            base($"Response {result} {errorMessage ?? ""}: {responsePayload}") {
+            base($"{{\"Message\":\"Response {result} {errorMessage ?? ""}\",\"Description\":{responsePayload}}}") {
```
Comment (karok2m) [inline common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs]: Please add a test for deserialization in the case when `responsePayload` is `null`.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/tests/Engine/StandaloneJobOrchestratorTests.cs [hunk 5]
```diff
@@ -124,19 +124,96 @@ public class StandaloneJobOrchestratorTests : TempFileProviderBase {
...
+            };
+            serializeExceptionModel = _newtonSoftJsonSerializer.SerializeToString(exceptionModel);
+            serializeExceptionModel.Should().BeEquivalentTo(exceptionResponse);
+        }
+
+            [Fact]
+        public async Task Test_PublishNodes_NullOrEmpty() {
+            InitStandaloneJobOrchestrator();
+
+            var exceptionResponse = $"{{\"Message\":\"Response 400 null request is provided\",\"Description\":{{}}}}";
```
Comment (karok2m) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/tests/Engine/StandaloneJobOrchestratorTests.cs]: `exceptionResponse` object does not seem to be used in `Test_PublishNodes_NullOrEmpty`. Can we remove it ?

File: common/src/Microsoft.Azure.IIoT.Hub.Module.Framework/src/Hosting/MethodRouter.cs (hunk 3 above)
Comment (dacolgit) [reply common/src/Microsoft.Azure.IIoT.Hub.Module.Framework/src/Hosting/MethodRouter.cs]: The exception response is not the same.
`throw new MethodCallStatusException(ex.Message, status);` generates:
`{
    "status":404,
    "payload":{
        "Message":"Response 404 Nodes not found",
        "Description":{
            "EndpointUrl":"opc.tcp://192.168.1.70:50000",
            "UseSecurity":false,
            "OpcAuthenticationMode":"Anonymous",
            "OpcNodes":[
                {
                    "Id":"ns=2;s=SlowUInt1",
                    "OpcPublishingInterval":1000
                }
            ]
        }
    }
}
`
while `throw ex` generates:
`{
    "status":404,
    "payload":{
        "EndpointUrl":"opc.tcp://192.168.1.70:50000",
        "UseSecurity":false,
        "OpcAuthenticationMode":"Anonymous",
        "OpcNodes":[
            {
                "Id":"ns=2;s=SlowUInt1",
                "OpcPublishingInterval":10000
            }
        ]
    }
}
`

Comment (karok2m) [reply common/src/Microsoft.Azure.IIoT.Hub.Module.Framework/src/Hosting/MethodRouter.cs]: @dacolgit, is this the result of direct method invocation ?

File: common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs (hunk 4 above)
Comment (dacolgit) [reply common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs]: added

File: common/src/Microsoft.Azure.IIoT.Hub.Module.Framework/src/Hosting/MethodRouter.cs (hunk 3 above)
Comment (karok2m) [reply common/src/Microsoft.Azure.IIoT.Hub.Module.Framework/src/Hosting/MethodRouter.cs]: I believe that this difference is caused by the fact that here you are transferring the value that has bees stored in `Message` to `ResponsePayload`, effectively overwriting it. And `ResponsePayload` is what is being picked up by `MethodRouter` when assembling the response to a direct method call. This makes me realize that we are incorrectly assigning assembled JSON to the `Message` field instead of `ResponsePayload` in the exception. So I think that we need to change the constructors that we have to contain the JSON message in `ResponsePayload`.

So this:

```csharp
        public MethodCallStatusException(int result, string errorMessage = null) :
            this("{}", result, errorMessage) {
            ResponsePayload = $"\"{errorMessage}\"";
        }

        ...
        public MethodCallStatusException(string responsePayload, int result,
            string errorMessage = null) :
            base($"{{\"Message\":\"Response {result} {errorMessage ?? ""}\",\"Description\":{responsePayload}}}") {
            Result = result;
            ResponsePayload = responsePayload;
        }
```

Should become something like this:

```csharp
        public MethodCallStatusException(int result, string errorMessage) :
            this("{}", result, errorMessage) { }

        ...
        public MethodCallStatusException(string responsePayload, int result, string errorMessage = null) :
            base($"Response {result} {errorMessage ?? ""}") {
            Result = result;
            ResponsePayload = $"{{\"Message\":\"Response {result} {errorMessage ?? ""}\",\"Description\":{responsePayload ?? "{}"}}}";
        }
```

File: common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs [hunk 6]
```diff
@@ -39,7 +39,7 @@ public class MethodCallStatusException : MethodCallException {
...
             Result = result;
             ResponsePayload = responsePayload;
```
Comment (karok2m) [inline common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs]: Please note that `ResponsePayload` in the first constructor does not generate a proper JSON. We need to change that as `ResponsePayload` is picked up by `MethodRouter` to be the payload of direct method call response. So we need to fix that. Please see my previous comment about storing JSON payload in `Message` vs `ResponsePayload`, as one of the options would be to not overwrite `ResponsePayload` in the first constructor after it is being assigned here and make sure that we have a proper JSON here.

File: common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs (hunk 4 above)
Comment (karok2m) [inline common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs]: [Microsoft REST API Guidelines](https://github.com/Microsoft/api-guidelines/blob/master/Guidelines.md#710-response-formats) recommend to use `Details` together with `Message`. So maybe we should rename `Description` to `Details` ?

A few more points from there:
* `Details` there is also a list of objects instead of a single object, but I think that we can keep our single object here.
* The top level object also contains `error` key with the whole object that we assemble here as its value. In our case the object will become the value for `payload` key. So I am happy with that part of our existing structure as well.

What do you think, @dacolgit, @cristipogacean ?

Comment (dacolgit) [reply common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs]: I'm ok for renaming `Description` to `Details`
@cristipogacean ?

File: common/src/Microsoft.Azure.IIoT.Hub.Module.Framework/src/Hosting/MethodRouter.cs (hunk 3 above)
Comment (dacolgit) [reply common/src/Microsoft.Azure.IIoT.Hub.Module.Framework/src/Hosting/MethodRouter.cs]: @dacolgit, is this the result of direct method invocation ?
Yes

File: common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs (hunk 6 above)
Comment (dacolgit) [reply common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs]: yes changed

File: common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs (hunk 4 above)
Comment (karok2m) [reply common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs]: Let's rename `Description` to `Details` then.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/tests/Utils/MethodCallStatusExeptionModel.cs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+        /// <summary>
+        /// Exception message.
+        /// </summary>
+        [DataMember(Name = "Message", Order = 0,
+            EmitDefaultValue = true)]
+        public string Message { get; set; }
+
+        /// <summary>
+        /// Description.
```
Comment (karok2m) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/tests/Utils/MethodCallStatusExeptionModel.cs]: Let's change `Description` to something like `Details of the exception.`.

Comment (dacolgit) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/tests/Utils/MethodCallStatusExeptionModel.cs]: changed

**Review by karok2m — DISMISSED** [review]

File: common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs [hunk 8]
```diff
@@ -38,10 +37,10 @@ public class MethodCallStatusException : MethodCallException {
         /// <param name="result"></param>
         /// <param name="errorMessage"></param>
         public MethodCallStatusException(string responsePayload, int result,
-            string errorMessage = null) :
-            base($"Response {result} {errorMessage ?? ""}: {responsePayload}") {
+           string errorMessage = null) :
```
Comment (cristipogacean) [inline common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs]: please fix indentation

File: common/src/Microsoft.Azure.IIoT.Hub.Module.Framework/src/Hosting/MethodRouter.cs [hunk 9]
```diff
@@ -300,6 +300,7 @@ private class DefaultFilter : ExceptionFilterAttribute {
                         }
                         _logger.Verbose(ex, "Method call error");
                         ex = _ef.Filter(ex, out var status);
+
```
Comment (cristipogacean) [inline common/src/Microsoft.Azure.IIoT.Hub.Module.Framework/src/Hosting/MethodRouter.cs]: let's revert line changes here

File: common/src/Microsoft.Azure.IIoT.Hub.Module.Framework/src/Hosting/MethodRouter.cs [hunk 10]
```diff
@@ -325,6 +326,7 @@ private class DefaultFilter : ExceptionFilterAttribute {
...
```
Comment (cristipogacean) [inline common/src/Microsoft.Azure.IIoT.Hub.Module.Framework/src/Hosting/MethodRouter.cs]: let's revert line changes here

File: common/src/Microsoft.Azure.IIoT.Serializers.NewtonSoft/src/Default/NewtonSoftJsonSerializer.cs [hunk 11]
```diff
@@ -534,6 +534,29 @@ internal sealed class JsonVariantConverter : JsonConverter {
             private readonly NewtonSoftJsonSerializer _serializer;
         }
 
+        /// <summary>
+        /// Json raw converter
+        /// </summary>
+        public sealed class RawJsonConverter : JsonConverter {
```
Comment (cristipogacean) [inline common/src/Microsoft.Azure.IIoT.Serializers.NewtonSoft/src/Default/NewtonSoftJsonSerializer.cs]: since this is only used in the tests, can we move it in a Tests related class? next to MethodCallStatusExceptionModel maybe?

Comment (dacolgit) [reply common/src/Microsoft.Azure.IIoT.Serializers.NewtonSoft/src/Default/NewtonSoftJsonSerializer.cs]: moved

File: common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs (hunk 8 above)
Comment (dacolgit) [reply common/src/Microsoft.Azure.IIoT.Abstractions/src/Exceptions/MethodCallStatusException.cs]: fixed

**Review by karok2m — APPROVED** [review]

**Review by cristipogacean — APPROVED** [review]
